### PR TITLE
XRDDEV-681 Subject Search

### PR DESCRIPTION
**Pull request by carohauta**
## Description

**It may be worth to wait for #101 to be merged before this one since there will be conflicts**

Adds Subject search so that adding access rights will be possible.

**THIS PR WILL BE CLOSED - DO NOT MERGE AT ALL**

PR #108 includes changes in this one

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have followed the agreed [Version Control Practices](https://confluence.niis.org/pages/viewpage.action?spaceKey=XRDDEV&title=Version+Control+Practices)
- [x] My changes generate no new warnings or errors (e.g. Javascript console, Java stdout)
- [x] I have made corresponding changes to the documentation
- [x] The new code has sufficient test coverage
- [x] The build, unit and integration tests pass
- [x] There is a link to a successful Jenkins build
- [x] No new npm audit issues, or new issues have been added to [Front-end build process](https://confluence.niis.org/display/XRDDEV/Front-end+build+process) tracking table and accepted
- [x] New backlog items that have been created (such as "not implementing this acceptance criteria now") are mentioned + linked in Jira task comments
- [x] All task outputs (PR, documentation, UI design, etc) is listed in a Jira comment so that it is easy for reviewer to check
- [x] Deviations from acceptance criteria listed in comments (also if "this criteria was removed")

### CODE REVIEW

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/ClientsApiController.java [hunk 1]
```diff
@@ -338,4 +352,22 @@ private ClientType getClientType(String encodedId) {
         return createCreatedResponse("/api/service-descriptions/{id}", addedServiceDescription,
                 addedServiceDescription.getId());
     }
+
+    @Override
+    @PreAuthorize("hasAuthority('VIEW_CLIENT_ACL_SUBJECTS')")
+    public ResponseEntity<List<Subject>> findSubjects(String encodedClientId, String memberNameGroupDescription,
```
[the remark this answers is not in the document]
Comment (carohauta) [reply src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/ClientsApiController.java]: Yeah this is pretty much derived from the current implementation and the fact that the current UI design supports this sort of combination parameter. I updated the name or the parameter: https://github.com/nordic-institute/X-Road-REST-UI/pull/108/files#diff-b448ad26221fc349d0a63e9fd69c1073R362

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/dto/AccessRightHolderDto.java [hunk 2]
```diff
@@ -36,17 +36,21 @@
...
+    private String groupId;
     /**
-     * LocalGroup's groupCode - NULL if not a LOCALGROUP
+     * Group's groupCode - NULL if not a LOCALGROUP/GLOBALGROUP
      */
-    private String localGroupCode;
+    private String groupCode;
     /**
-     * LocalGroup's groupDescription - NULL if not a LOCALGROUP
+     * Group's groupDescription - NULL if not a LOCALGROUP/GLOBALGROUP
```
[the remark this answers is not in the document]
Comment (carohauta) [reply src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/dto/AccessRightHolderDto.java]: Renamed: https://github.com/nordic-institute/X-Road-REST-UI/pull/108/files#diff-821e42f4ad29fcbae7c5bd86ff91f6dfR47

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/dto/AccessRightHolderDto.java [hunk 3]
```diff
@@ -36,17 +36,21 @@
...
+     * Group's PK - NULL if not a LOCALGROUP/GLOBALGROUP
      */
-    private String localGroupId;
+    private String groupId;
     /**
-     * LocalGroup's groupCode - NULL if not a LOCALGROUP
+     * Group's groupCode - NULL if not a LOCALGROUP/GLOBALGROUP
      */
-    private String localGroupCode;
+    private String groupCode;
```
[the remark this answers is not in the document]
Comment (carohauta) [reply src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/dto/AccessRightHolderDto.java]: Renamed: https://github.com/nordic-institute/X-Road-REST-UI/pull/108/files#diff-821e42f4ad29fcbae7c5bd86ff91f6dfR43

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/dto/AccessRightHolderDto.java [hunk 4]
```diff
@@ -36,17 +36,21 @@
 @Data
 public class AccessRightHolderDto {
     /**
-     * LocalGroup's PK - NULL if not a LOCALGROUP
+     * Group's PK - NULL if not a LOCALGROUP/GLOBALGROUP
```
[the remark this answers is not in the document]
Comment (carohauta) [reply src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/dto/AccessRightHolderDto.java]: Renamed: https://github.com/nordic-institute/X-Road-REST-UI/pull/108/files#diff-821e42f4ad29fcbae7c5bd86ff91f6dfR41

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ServiceService.java [hunk 5]
```diff
@@ -306,9 +317,165 @@ public ServiceNotFoundException(String s) {
...
         public static final String ERROR_ACCESSRIGHT_NOT_FOUND = "accessright_not_found";
+
         public AccessRightNotFoundException() {
             super(new ErrorDeviation(ERROR_ACCESSRIGHT_NOT_FOUND));
         }
     }
 
+    /**
+     * Find access right holders by search terms
+     * @param clientId
```
[the remark this answers is not in the document]
Comment (carohauta) [reply src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ServiceService.java]: Added this to javadoc: https://github.com/nordic-institute/X-Road-REST-UI/pull/108/files#diff-6ee6f01b69867a70e548934bb1440dfeR329

**Comment by carohauta** [comment]
Closed this PR since #108 will include everything in this one and then some

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ServiceService.java [hunk 6]
```diff
@@ -306,9 +317,165 @@ public ServiceNotFoundException(String s) {
...
+        if (!StringUtils.isEmpty(subsystemCode)) {
+            searchPredicate = searchPredicate.and(dto -> {
+                XRoadId xRoadId = dto.getSubjectId();
+                if (xRoadId instanceof ClientId) {
+                    String clientSubsystemCode = ((ClientId) xRoadId).getSubsystemCode();
+                    return clientSubsystemCode != null && clientSubsystemCode
+                            .toLowerCase()
+                            .contains(subsystemCode.toLowerCase());
+                } else {
+                    return false;
```
[the remark this answers is not in the document]
Comment (jansu76) [reply src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ServiceService.java]: pretty high class comments there 💪

**Comment by jansu76** [comment]
Reviewed fixes to my comments, all good.

**Pull request closed without merging**